optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java

```diff
@@ -0,0 +1,371 @@
+
+import org.optaplanner.core.api.score.buildin.hardsoft.HardSoftScore;
+import org.optaplanner.core.api.score.stream.Constraint;
+import org.optaplanner.core.api.score.stream.ConstraintCollectors;
+import org.optaplanner.core.api.score.stream.ConstraintFactory;
+import org.optaplanner.core.api.score.stream.ConstraintProvider;
+import org.optaplanner.examples.examination.domain.Exam;
+import org.optaplanner.examples.examination.domain.Period;
+import org.optaplanner.examples.examination.domain.PeriodPenalty;
+import org.optaplanner.examples.examination.domain.PeriodPenaltyType;
+import org.optaplanner.examples.examination.domain.Room;
+import org.optaplanner.examples.examination.domain.RoomPenalty;
+import org.optaplanner.examples.examination.domain.RoomPenaltyType;
+import org.optaplanner.examples.examination.domain.solver.TopicConflict;
+
+import static org.optaplanner.core.api.score.stream.Joiners.equal;
+import static org.optaplanner.core.api.score.stream.Joiners.filtering;
+import static org.optaplanner.core.api.score.stream.Joiners.greaterThan;
+import static org.optaplanner.core.api.score.stream.Joiners.lessThan;
+import static org.optaplanner.core.api.score.stream.Joiners.lessThanOrEqual;
+
+public class ExaminationConstraintProvider implements ConstraintProvider {
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[]{
+                conflictingExamsInSamePeriod(constraintFactory),
+                periodDurationTooShort(constraintFactory),
+                roomCapacityTooSmall(constraintFactory),
+                periodPenaltyExamCoincidence(constraintFactory),
+                periodPenaltyExclusion(constraintFactory),
+                periodPenaltyAfter(constraintFactory),
+                roomPenaltyExclusive(constraintFactory),
+                twoExamsInARow(constraintFactory),
+                twoExamsInADay(constraintFactory),
+                periodSpread(constraintFactory),
+                mixedDurations(constraintFactory),
+                frontLoad(constraintFactory),
+                periodPenalty(constraintFactory),
+                roomPenalty(constraintFactory)
+        };
+    }
+
+    //// Two exams in the same period which share students.
+//    rule "conflictingExamsInSamePeriod"
+//    when
+//    $topicConflict : TopicConflict($leftTopic : leftTopic, $rightTopic : rightTopic)
+//    $leftExam : Exam(topic == $leftTopic, $period : period, period != null)
+//    $rightExam : Exam(topic == $rightTopic, period == $period)
+//    then
+//        scoreHolder.addHardConstraintMatch(kcontext, - $topicConflict.getStudentSize());
+//    end
+    private Constraint conflictingExamsInSamePeriod(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Exam.class)
+                .filter(exam -> exam.getPeriod() != null)
+                .join(Exam.class, equal(Exam::getPeriod, Exam::getPeriod))
```

Comment:
Also, these two joins can be replaced by `.fromUniquePair(Exam.class, equal(...))`